app/services/retrieve_pipeline_viz_graph_data_service.rb

```diff
@@ -0,0 +1,157 @@
+class RetrievePipelineVizGraphDataService
```

Comment:
Would it make sense to have tests for this service?

Reply:
We could, but they would look pretty similar to the tests for the pipeline_viz_controller, which essentially tests the outcome of this service along with some permission-ing.